### PR TITLE
enhance structure reuse and some minor optim (fixes #78)

### LINKED ISSUE #78: Monocle should not create new useless data structure
For instance when a modify returns the element it was sent.
@gcanti I may provide a PR with those optimisation if accepted beforehand.

### PR DESCRIPTION
**Pull request by sledorze**
@gcanti we can discuss those changes

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -143,11 +143,14 @@ function lensFromPath(path: Array<any>): any {
 }
 
 function lensFromProp<S, P extends keyof S>(prop: P): Lens<S, S[P]> {
-  return new Lens(s => s[prop], a => s => Object.assign({}, s, { [prop as any]: a }))
+  return new Lens(s => s[prop], a => s => (s[prop] === a ? s : Object.assign({}, s, { [prop as any]: a })))
```
Comment (gcanti) [inline src/index.ts]: maybe a little helper here?

```ts
const update = <O, K extends keyof O, A extends O[K]>(o: O, k: K, a: A): O => {
  return a === o[k] ? o : { ...o, [k]: a }
}

function lensFromProp<S, P extends keyof S>(prop: P): Lens<S, S[P]> {
  return new Lens(s => s[prop], a => s => update(s, prop, a))
}

function lensFromNullableProp<S, A extends S[K], K extends keyof S>(k: K, defaultValue: A): Lens<S, A> {
  return new Lens((s: any) => fromNullable(s[k]).getOrElse(defaultValue), a => s => update(s, k, a))
}

...

function optionalFromNullableProp<S, K extends keyof S>(k: K): Optional<S, NonNullable<S[K]>> {
  return new Optional((s: any) => fromNullable(s[k]), a => s => update(s, k, a))
}
```

**Comment by gcanti** [comment]
There's another opportunity at line ~780

```diff
  /** compose a Setter with a Setter */
  compose<B>(ab: Setter<A, B>): Setter<S, B> {
-    return new Setter(f => s => this.modify((a: A) => ab.modify(f)(a))(s))
+    return new Setter(f => s => this.modify(ab.modify(f))(s))
  }
```

**Comment by gcanti** [comment]
Actually

```ts
  /** compose a Setter with a Setter */
  compose<B>(ab: Setter<A, B>): Setter<S, B> {
    return new Setter(f => this.modify(ab.modify(f)))
  }
```

right?

**Comment by sledorze** [comment]
@gcanti still cannot run the test locally; I've done several improvements.
Finger crossed with the CI.

**Comment by gcanti** [comment]
Oh, my fault. I guess should be

```ts
const update = <O, K extends keyof O, A extends O[K]>(o: O, k: K, a: A): O => {
  return a === o[k] ? o : Object.assign({}, o, { [k]: a })
}
```

**Comment by sledorze** [comment]
@gcanti what do we do now?

**Comment by gcanti** [comment]
@sledorze add some tests to avoid regressions?

**Comment by sledorze** [comment]
I'm having a hard time not being able to run test locally, will try to understand why..

**Comment by sledorze** [comment]
Just in case you have an idea:

```> monocle-ts@1.6.0 dtslint /Users/sledorze/projects/monocle-ts
> dtslint dtslint

Running type-level tests for the following versions: ["3.0","3.1","3.2","3.3"]
AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert(maxIdx >= minIdx)

    at range (/Users/sledorze/projects/monocle-ts/node_modules/dtslint/bin/lint.js:168:5)
    at /Users/sledorze/projects/monocle-ts/node_modules/dtslint/bin/lint.js:146:36
    at Generator.next (<anonymous>)
    at fulfilled (/Users/sledorze/projects/monocle-ts/node_modules/dtslint/bin/lint.js:4:58)
npm ERR! code ELIFECYCLE```

**Comment by sledorze** [comment]
ok a clean of node_modules solved the problem..
I'll check

**Comment by sledorze** [comment]
@gcanti I think I've covered the functions using 'update'
Good to go?

File: src/index.ts [hunk 2]
```diff
@@ -15,6 +15,10 @@ import { identity, constant, Predicate, Refinement } from 'fp-ts/lib/function'
 import { identity as id } from 'fp-ts/lib/Identity'
 import { Const, getApplicative } from 'fp-ts/lib/Const'
 
+export const update = <O, K extends keyof O, A extends O[K]>(o: O, k: K, a: A): O => {
```
Comment (gcanti) [inline src/index.ts]: This is exported only for testing purposes right?

Let's tag it as internal

```ts
/**
 * @internal
 */
```

**Pull request merged**